Feature/content search

nayoonk928 opened the pull request:
### 변경사항
**AS-IS**

* 컨텐츠 검색 기능 구현
  * TmdbApiClinet 클래스를 통해  API 에서 검색 결과 반환
  * ContentService 클래스의 searchContents 메서드를 통해 API 값을 파싱하여 ContentSearchDto 객체의 리스트 반환
  * 검색 결과가 없으면 CustomExcception 호출
  * RedisCacheConfig 클래스를 통해 검색한 데이터 캐싱 처리

* ContentService 테스트 코드 작성

**TO-BE**

* 컨텐츠 상세정보 보기 구현

### 테스트
- [x] 테스트 코드
- [x] API 테스트

nayoonk928 committed:
feat: Implement content search feature
- 제목, 배우, 감독으로 컨텐츠를 검색

nayoonk928 committed:
refactor: Change id type from int to long

nayoonk928 committed:
feat: Add caching of search results to Redis
- 컨텐츠 검색 결과 레디스 캐시에 저장하는 기능 추가

nayoonk928 committed:
refactor: Add error handling
- 컨텐츠 검색시 결과가 없을 때 NO_RESULTS_FOUND 반환
- 컨텐츠 검색 테스트 코드 작성

mingming-mentor commented on src/main/java/com/personal/contentservice/config/GenreDataInitializer.java:
```diff
@@ -0,0 +1,62 @@
+      }
+
+    } catch (IOException e) {
+      e.printStackTrace();
```
log를 이용해서 에러 남겨주시면 좋을 것 같아요

mingming-mentor commented on src/main/java/com/personal/contentservice/config/TmdbApiClient.java:
```diff
@@ -0,0 +1,49 @@
+public class TmdbApiClient {
+
+  private final HttpClient httpClient;
+  private static final String TMDB_API_BASE = "https://api.themoviedb.org/3/";
```
요 api base도 application.yaml(application.properties)로 빼주셔도 좋을 것 같아요.
실무에서는 실 서비스에 나가기전 여러 phase를  두고, 이 안에서 테스팅을 하는데요.
(예시 : alpha, beta, stage, real 식으로 phase를 두고 각 단계별 테스팅을 진행합니다. real이 실서비스)

내부 서비스끼리 통신을 할때 각 phase에 맞춰서 통신을 하게 됩니다. 따라서 host 주소가 항상같지가 않습니다. 
그래서 보통 api host주소를 config에 저장하여 사용하게 되어요

mingming-mentor replied:
이 부분은 현재유효한 부분은 아니니 참고용으로만 생각하셔도 좋을 것 같아요

mingming-mentor commented on src/main/java/com/personal/contentservice/config/TmdbApiClient.java:
```diff
@@ -0,0 +1,49 @@
+    return response.body();
+  }
+
+  public String searchContents(String query, int page) throws Exception {
```
Tmdb에서 제공하는 API가 restful api인것 같은데 맞을까요?
getString을 이용해 string으로 가지고 오지 마시고, response에 맞춰 class를 정의하시면 보다 편하게 데이터를 다룰 수 있을 듯 합니다. 
https://velog.io/@seongwon97/Spring-Boot-Rest-Template (restTemplate을 활용한거긴 하지만, 이런식으로 사용하시면 될것 같아요)

mingming-mentor commented on src/main/java/com/personal/contentservice/controller/ContentController.java:
```diff
@@ -0,0 +1,24 @@
+  private final ContentService contentService;
+
+  @GetMapping("/search/{query}/{page}")
+  public ResponseEntity<?> searchContents(
```
query는 보통 queryParam으로 많이 사용합니다. https://restfulapi.net/resource-naming/ 2.5번 참고해주세요

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ContentServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+    return contentSearchDtoList;
+  }
+
+  private ContentSearchDto extractContent(JsonNode result, DateTimeFormatter formatter) {
```
이 부분들도 httpClient 응답을 reponse에 맞는 class로 받으면 많이 사라지게 될듯해요. 훨씬 가독성이 좋게 될듯합니다

mingming-mentor reviewed (approved):
리뷰드렸습니다! 수고 많으셨어요

nayoonk928 committed:
refactor: Refactor search content functionality

- MediaTypeDto 및 그 하위 클래스 추가
- 장르 ID를 제공된 장르 Map 을 사용하여 genreName으로 변환
- MovieSearchDto 및 TvSearchDto에 대한 별도의 Response 클래스 추가
- ContentServiceImpl 수정
- TmdbApiClient 에서 SearchResponseDto를 반환하도록 수정

nayoonk928 committed:
test: Add ContentServiceImplTest class

nayoonk928 commented:
@mingming-mentor 
말씀해 주신 사항들 반영하여 코드 수정하였습니다. 더 수정해야 하는 부분 있는지 리뷰 부탁드립니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ContentServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    SearchResponseDto response = tmdbApiClient.searchContents(query, page);
+
+    // 현재 시간 이후에 공개되는 작품은 제외
+    List<MediaTypeDto> results = response.getResults().stream()
+        .filter(result -> {
+          if (result instanceof MovieSearchDto) {
+            MovieSearchDto movie = (MovieSearchDto) result;
+            return isDateBeforeNow(movie.getDate());
+          } else if (result instanceof TvSearchDto) {
+            TvSearchDto tvShow = (TvSearchDto) result;
+            return isDateBeforeNow(tvShow.getDate());
+          } else {
+            return true;
+          }
+        })
+        .collect(Collectors.toList());
```
date라는 필드가 MovieSearchDto, TvSearchDto에도 있는 필드라면 MediaTypeDto에 두어도 될것 같아요. 그럼 converting하는 로직을 줄일 수 있을 듯 합니다

nayoonk928 replied:
api, service dto를 분리하면서 if문 하나로 통합 했습니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/GenreServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+  private final GenreRepository genreRepository;
+
+  public Map<Long, String> getAllGenres() {
+    List<Genre> genres = genreRepository.findAll();
+    Map<Long, String> genreMap = new HashMap<>();
+    for (Genre genre : genres) {
+      genreMap.put(genre.getId(), genre.getName());
+    }
+    return genreMap;
```
```
Map<Long, String> genreMap = genreRepository.findAll().stream().collect(Collectors.toMap(it -> it.getId(), it -> it.getName()));
```
로 간단히 줄여볼 수 있을 듯 합니다.

nayoonk928 replied:
감사합니다! 수정했습니다.

mingming-mentor commented on src/main/java/com/personal/contentservice/dto/search/MovieSearchDto.java:
```diff
@@ -0,0 +1,67 @@
+  private long id;
+  private String title;
+
+  @JsonProperty("media_type")
```
https://aandi.tistory.com/24 요 블로그를 참고해보시면 좀더 간단히 camelToSnake로 설정이 가능할 듯 해요. 
그리고 가능하면 TmdbApiClient에서 사용하는 dto와 service에서 사용하는 dto도 분리한뒤, 외부 api통신에서 사용하는 dto는 XXResponse라는 이름으로 설정하면 좀더 layer 분리를 명확히 할 수 있을 듯 해요. 이 부분은 참고로 봐주세요!

nayoonk928 replied:
api, service 용 dto로 분리했습니다.

mingming-mentor reviewed (approved):
리뷰드렸습니다!

nayoonk928 committed:
refactor: Refactor search DTOs and service

- api통신 dto와 service dto 분리
- dto에 맞춰 service 코드와 test 코드 수정

JaeYooooon commented on src/main/java/com/personal/contentservice/config/RedisCacheConfig.java:
레디스 캐싱에 대해 공부하고갑니다.
